### ISSUE TITLE
Add prosopite

### ISSUE THREAD

**Issue by pushcx** [lobsters/lobsters#1252]
Add [https://github.com/charkost/prosopite](prosopite) to detect 1+n queries.  Configure it to raise in dev/test and use the rails logger in prod (I’ll probably turn this off in prod if nothing shows up in a few weeks).

**Comment by mikalai-yankouski** [comment]
please take a look

**Comment by pushcx** [comment]
Merged #1260 by @mikalai-yankouski to complete this.

**Comment by byroot** [comment]
> (I’ll probably turn this off in prod if nothing shows up in a few weeks).

FYI: It's still active in production, and adds a non-negligible overhead, not sure if it was intended.

**Comment by pushcx** [comment]
We don't have 100% code path test coverage so the production logging is the only way to find some 1 + n queries. I've gone and looked at the log a few times to fix them, but it's been a little while since the last attempt. I figured at some point there will be few enough that I'll set it to throw instead of log in prod, but I haven't gotten there. 

I haven't found a tool or practice that reliably prevents ActiveRecord from producing more of them. Suggestions would be very welcome.

**Comment by byroot** [comment]
I think the vanilla way is to use `strict_loading`: https://api.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-strict_loading

**Comment by pushcx** [comment]
I can see how that's a start, but I don't see how to enforce its use. A rubocop rule would have to catch any mention of a model followed by a method that touches the broad and growing AR query API, and then also check the models for `self.` invocations, and I'm sure there's more I'm missing.

**Comment by byroot** [comment]
> I don't see how to enforce its use.

Via default scopes? e.g. to enable globally:

```ruby
class ApplicationRecord
  default_scope { strict_loading }
end
```

But can also be done gradually on a per class basis.

**Comment by pushcx** [comment]
Huh. I'm so used to `default_scope` as a footgun that it didn't occur to me.

I regularly walk associations in the console for development and debugging, so I started writing an initializer to remove the `default_scopes`.*

That worked, but the error message and traceback for `strict_loading` and `StrictLoadingViolationError` are pretty confusing. They don't lead to where it's configured so I can leave an explanation, and the top google hits are not great. A lot of Lobsters contributors are new to Rails and I've seen them get stuck on things like this.

As I was looking around for how to improve that, I ran into an undocumented option `strict_loading_by_default` that I can set per-environment to recreate the Prosopite behavior of logging in prod and raising in dev and test. I'll try setting this up on Monday.


 \* that unused initializer:
```
ActiveSupport::Reloader.to_prepare do
  if Rails.const_defined?(:Console)
    # list models as the get added, can be enforced with a cop
    [User, Story].each do |model|
      model.default_scopes = []
    end
  end
end
```

**Comment by pushcx** [comment]
I spent a few minutes on today's stream dong an experiment to remove Prosopite and enable `strict_loading` in dev and test envs. Unfortunately it doesn't look like we could remove the overhead of Prosopite. `strict_loading` doesn't just prohibit 1 + n queries but using associations in any way except when preloaded. We do that all the time! So we'd have to do a large refactoring to every entry point (controllers, jobs, scripts, etc) that eventually loads records from the database. I guess that might be possible if we used `default_scope` to go model-by-model and eventually replaced it with `strict_loading`, but that feels like a ton of work to get slightly worse 1 + n detection. (I say slightly worse because strict_loading doesn't catch `3.times { |i| User.find(i) }`; prosopite has.)
